File: Programmers/C30L42897/C30L42897.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

int solution(vector<int> money) {
    //
    vector<int> caching0(money.size()+1, 0);
    caching0[money.size() - 2] = money[money.size() - 2];
    // caching0[money.size() - 1] = 0;
    for (int cur = money.size() - 3; cur >= 0; cur--) {
        caching0[cur] = money[cur] + max(caching0[cur + 2], caching0[cur + 3]);
    }
    //
    vector<int> caching1(money.size() + 2, 0);
    caching1[money.size() - 1] = money[money.size() - 1];
    // caching1[money.size()] = 0;
    for (int cur = money.size() - 2; cur >= 1; cur--) {
        caching1[cur] = money[cur] + max(caching1[cur + 2], caching1[cur + 3]);
    }

    int a = max({caching0[0], caching0[1], caching1[1], caching1[2]});
    return a;
}

// int main() { cout << solution({1, 2, 3, 1}) << endl; }
int main() {
    // cout << solution({123, 230, 312, 125, 123, 127, 129}) << endl;
    // cout << solution({1000, 0, 0, 0, 0, 1000, 0, 0, 0, 0, 0, 1000}) << endl;
    // cout << solution({1,2,3,4,5,6,7,8,9,10}) << endl;
    // cout << solution({11,0,2,5,100,100,85,1}) << endl;
    // cout << solution({91,90,5,7,5,7}) << endl;
    cout << solution({90, 0, 0, 95, 1, 1}) << endl;
    // cout << solution({90, 0, 0, 0, 95, 1, 1}) << endl;
    // cout << solution({0,0,0,95,1,90}) << endl;
}

File: Programmers/C30L42897/C30L42897.py
# %%
import functools as ft
import timeit


def solution(money):
    @ft.lru_cache(maxsize=None)
    def calc0(start):
        if start > len(money) - 2:
            return 0
        if start == len(money) - 2:
            return money[start]
        return max(money[cur] + calc0(cur + 2) for cur in range(len(money) - 2, start-1, -1))

    @ft.lru_cache(maxsize=None)
    def calc1(start):
        if start > len(money) - 1:
            return 0
        if start == len(money) - 1:
            return money[start]
        return max(money[cur] + calc1(cur + 2) for cur in range(len(money) - 1, start-1, -1))
    return max(
        money[0] + calc0(2),
        calc1(1)
    )


# def solution2(money):
#     c0 = [(2, money[0])]
#     l0 = len(money) - 1
#     c1 = [(1, money[1])]
#     l1 = len(money) - 1
#     return max(
#         money[0] + calc0(2),
#         calc1(1)
#     )


# print(solution([i for i in range(30000)]))
# print(solution([i for i in range(10000)]))
# print(solution([i for i in range(9000)]))
# print(solution([i for i in range(8000)]))
# print(solution([i for i in range(7000)]))
# print(solution([i for i in range(6000)]))
# print(timeit.timeit(lambda: solution([i for i in range(3000)]), number=1))
# print(timeit.timeit(lambda: solution([i for i in range(6000)]), number=1))
# print(solution([i for i in range(5000)]))
# print(solution([i for i in range(3000)]))
# print(solution([i for i in range(2000)]))
# print(solution([i for i in range(1000)]))
print(solution([
    123, 230, 312, 125, 123, 127, 129
]))

# %% 첫번째 시도


def solution(money):
    @ft.lru_cache(maxsize=None)
    def calc(start, end):
        if start > end:
            return 0
        if start == end:
            return money[start]
        db = []
        for cur in range(start, end + 1):
            v = calc(cur + 2, (end - 1) if cur == 0 else end)
            db.append(money[cur] + v)
        return max(db)
    for i in range(1, len(money)-1):
        calc(len(money)-1 - i, len(money)-1)
    return calc(0, len(money)-1)

# 방법은 맞으나 크기가 커지면 recursive overflow 발생
